File: src/1339_maximum_product_of_splitted_binary_tree/MaximumProductOfSplittedBinaryTree.java
// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Solution class for logic
class Solution {
    int MOD;
    long maxProduct;
    long totalSum;

    public int maxProduct(TreeNode root) {
        MOD = 1_000_000_007;

        totalSum = totalSum(root);
        dfs(root);

        return (int) (maxProduct % MOD);
    }

    private int totalSum(TreeNode root) {
        if (root == null)
            return 0;

        return totalSum(root.left) + totalSum(root.right) + root.val;
    }

    private long dfs(TreeNode root) {
        if (root == null) {
            return 0;
        }

        long subtreeSum = dfs(root.left) + dfs(root.right) + root.val;

        maxProduct = Math.max(maxProduct, (long) ((totalSum - subtreeSum) * subtreeSum));

        return subtreeSum;
    }
}

// ...existing code...

File: src/1339_maximum_product_of_splitted_binary_tree/maximum_product_of_splitted_binary_tree.cpp
#include <cstddef>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
   public:
    int maxProduct(TreeNode* root) {
        int MOD = 1'000'000'007;
        totalSum = treeSum(root);
        maxProd = 0;
        dfs(root);
        return (int)(maxProd % MOD);
    }

   private:
    long long maxProd = 0;
    int totalSum = 0;
    int treeSum(TreeNode* node) {
        if (!node) return 0;
        return node->val + treeSum(node->left) + treeSum(node->right);
    }
    int dfs(TreeNode* node) {
        if (!node) return 0;
        int subSum = dfs(node->left) + dfs(node->right) + node->val;
        maxProd = std::max(maxProd, 1LL * subSum * (totalSum - subSum));
        return subSum;
    }
};

// ...existing code...
